### ISSUE TITLE
"user cancelled" error on Android

### ISSUE THREAD

**Issue by zigzag-way** [EddyVerbruggen/cordova-plugin-googleplus#165]
When I`m using your plugin on my Nexus 4 it always return "user cancelled" error as a result of login (error callback). But it works fine on my Xiami phone.
Here is the code:
 $window.plugins.googleplus.login(
                                { offline: 'true', webApiKey: WEB_API_KEY },
                                function (user) {
                                    serverAuth(user);
                                },
                                function (msg) {
                                    $cordovaDialogs.alert('Google Plus login error. Please try again later', 'Login Error', 'OK');
                                }
                            );

**Comment by zigzag-way** [comment]
Looks similar to https://github.com/EddyVerbruggen/cordova-plugin-googleplus/issues/5

**Comment by sammuggPS** [comment]
Hey @zigzag-way,

I just merged a significant PR the other day that updated the Android side to use Google Sign-In and refactored plugin usage. The current documentation is up to date with those changes. You should check it out. To install it, use the second (GitHub URL) method found in the [README](https://github.com/EddyVerbruggen/cordova-plugin-googleplus#4-installation-phonegap-cli--cordova-cli).

Let me know if you have any questions!

Happy Coding,
~Sam
P.S. If your issue gets resolved, please close it!

**Comment by marianoarga** [comment]
Now I updated plugin and "google repository" from Android SDK manager it always returns "12501" error wich is the same (user cancelled).

**Comment by sammuggPS** [comment]
@marianoarga did you install the plugin from GitHub? This issue should have been resolved in the latest code (which _is not_ on NPM yet).

**Comment by marianoarga** [comment]
@sammuggPS Yes, and followed the "keytool SHA1" steps, since they where not needed before, used android-debug.apk in google console, used the SHA1 keytool gave me...
The only thing I removed was this line from installation command.
`-variable REVERSED_CLIENT_ID=myreversedclientid`
Even the demo app is returning the same 12501 error.

**Comment by marianoarga** [comment]
@zigzag-way Do you have any updates, did it work after updating the plugin?

**Comment by sammuggPS** [comment]
@marianoarga The demo app has not been updated... sorry. That's not something I got around to doing.

The reversed clientId is only needed for iOS. If you are only developing for Android, you don't need to include it.

**Comment by marianoarga** [comment]
@sammuggPS  Yep, that's why I did not include that on the install line, I'll be keep on trying and looking foward to see what happen to @zigzag-way project.
Thank you, really apreciate your efforts.

**Comment by sammuggPS** [comment]
@zigzag-way and @marianoarga,

I was just looking at another thread I was working on, for issue #190, and was reminded of one reason why you could be getting the 12501 issue:

_The Client Id you pass in to the login and trySilentLogin needs to be from the WEB app credentials of your project, not the Android client id._

This is very important! Please try it out and let me know what happens.

**Comment by marianoarga** [comment]
@sammuggPS I think i'm using the correct ones, the ones you see in this URL (having selected the project previously)
https://console.developers.google.com/apis/credentials/oauthclient

You may see the Google console like I see it here:
http://www.awesomescreenshot.com/image/1140931/022d1da815bcf2aa8c78e843378f8584

**Comment by sammuggPS** [comment]
@marianoarga, That is not the clientId you want. You need the one for your _WEB_ application, not Android.
![screen shot 2016-04-05 at 2 17 12 pm](https://cloud.githubusercontent.com/assets/8354240/14293272/e3f5c09e-fb39-11e5-9a73-92f7b188d254.png)

**Comment by marianoarga** [comment]
Ok, I'll give it a try, but I don't really understand why I need to submit the keytool's SHA1, since I'll not use that key from the Android app credentials...
Wil keep it up and let yu know.

**Comment by sammuggPS** [comment]
Think about it this way: A Cordova or Phonegap application has two parts, native and hybrid.

This plugin let's you authenticate using the native side. To authenticate a native Android App with Google you have to set up the app credentials. Android apps are recognized using the SHA1 fingerprint, which gets encoded into the installed package.

On the other hand, the hybrid side is running in a webview on top of a native application. Therefore, it is somewhat decoupled from the native side. If you need an access token and/or refresh token you should treat the webview as if it were a server, using a serverAuthCode to get those tokens to use for that session.

Remember, the webview doesn't persist data the same way that the native side does, so those tokens aren't going to be there the next time you start the app up. This is why we take advantage of the decoupled native and hybrid code. We let the native code manage the Google user and use the info we get back from it to do stuff in the hybrid code.

I _strongly_ urge you to dive into Google's Sign-In documentation and learn about all this. They will explain it much better than I, and you can see where I'm getting all these concepts from!

**Comment by marianoarga** [comment]
I see, I understand that now.
I have been testing with the same results (12501)

I'm using the webclient key.
http://www.awesomescreenshot.com/image/1141231/e01963d86fc65e50908ddca3ca4f1dec

This is the portion of code I'm using, it fires automatically after "deviceready"
```
function login() {
    window.plugins.googleplus.login(
            {
                 'scopes': 'profile',
                 'offline': false,
                 'webClientId': 'xxxxxxxxxxxxxxxxx.apps.googleusercontent.com'
            },
            function (obj) {
                console.debug(obj);
            },
            function (msg) {;
                console.debug(msg);
            }
    );
}
```

I have added the android-debug.apk debug cert (~/.android/debug.keystore) SHA1 to google console in order to use that debug certificate, I have tried the production one too with the same results.
http://www.awesomescreenshot.com/image/1141323/1bfed0e59d64e69a42f5912d1f18e38b

**Comment by SidharthPHariharan** [comment]
@sammuggPS So you are saying that we should use web credentials and not the native android credentials? The reason I am asking is that I am using web client id and still getting the error..

**Comment by SidharthPHariharan** [comment]
@marianoarga any luck.. did you make it work?
@zigzag-way You have been silent for some time did you have any luck with this plugin?

**Comment by marianoarga** [comment]
@SidharthPHariharan No, I didn't

**Comment by sammuggPS** [comment]
@SidharthPHariharan, Yes you need to use the web credentials not the native android credentials. This is specified in Google's documentation for getting an [`idToken`](https://developers.google.com/identity/sign-in/android/backend-auth) and [`serverAuthCode`](https://developers.google.com/identity/sign-in/android/offline-access).

**Comment by getcarpool** [comment]
Hello, 
First of all, thanks for writing this wonderful plugin.
I can confirm that this is in fact an issue but only with the latest GIT version of sourcecode.
To make things more clear, here's what I did:
1) I correctly used sha1 generated from keytool
2) I used webkey for webClientId
3) I used the latest version and got error code 12501
4) cordova plugin remove cordova-plugin-googleplus
5) cordova plugin add cordova-plugin-googleplus (NPM version)
6) Now I am able to see the google permissions dialog

can you please look into the latest fixes?
thanks

**Comment by SidharthPHariharan** [comment]
@marianoarga very sad..
@sammuggPS  great.. this is what I am using

`window.plugins.googleplus.login(
        {'scopes':[ 'userinfo_profile']},
        function (user_data) {},function(msg){
// this is where the control arrives after selecting the google account from popup
});`

why am I getting user cancelled even now? I am using the cli installation. The git version is clashing with existing plugins..

**Comment by sammuggPS** [comment]
@SidharthPHariharan, The only thing I can see is that you are not passing in your scopes properly. They need to be passed in as a space separated string. For example, `scopes: 'profile email'`.

We are aware of the git version clashing with other plugins...And we're not completely sure why. Is it the phonegap facebook plugin? or cordova-plugin-facebook? If so, you're not alone.

**Comment by SidharthPHariharan** [comment]
@sammuggPS changing the scope did not work as well. The clash was with google play services, I know its deprecated but another plugin requires it so its unavoidable.
Please help me with this asap.

**Comment by marianoarga** [comment]
@SidharthPHariharan Please avoid using words like ASAP in a collaborative community.

Just for adding some data, the list of plugins I'm using:

> cordova-plugin-console 1.0.2 "Console"
> cordova-plugin-contacts 2.0.1 "Contacts"
> cordova-plugin-device 1.1.1 "Device"
> cordova-plugin-fullscreen 1.1.0 "cordova-plugin-fullscreen"
> cordova-plugin-globalization 1.0.3 "Globalization"
> cordova-plugin-google-analytics 0.8.1 "Google Universal Analytics Plugin"
> cordova-plugin-googleplus 4.0.9 "Google+"
> cordova-plugin-sim 1.2.1 "SIM"
> cordova-plugin-whitelist 1.2.1 "Whitelist"
> phonegap-plugin-push 1.6.1 "PushPlugin"

**Comment by SidharthPHariharan** [comment]
@marianoarga ok(for not using asap as you have received some appreciation there :+1: )
but as I said another plugin requires google play services plugin and thats causing the problem with the git version of this plugin.

**Comment by getcarpool** [comment]
@sammuggPS , Any luck investigating this issue? I have been cracking my head since yesterday but no luck. 
Happen to visit this S/O url
http://stackoverflow.com/questions/33583326/new-google-sign-in-android
it says about using 2 API keys 
To quote:
**_Android client key will go inside the google-services.json, and the web application key needs to be set in the requestIdToken or requestServerAuthCode methods._**

Am i suppose to follow this?

**Comment by sammuggPS** [comment]
@getcarpool, Yes, you need to pass the Web App client ID, not the Android or iOS client IDs, in to the googleplus.login function.

Also, an API Key is a different value than the Client ID. You need to pass in the Client ID so the plugin can request an idToken and serverAuthCode.

**Comment by getcarpool** [comment]
@sammuggPS Thanks for your response. But I am afraid, that's exactly what I am passing. 
I am pretty sure I am following each and every step to the letter. Still no luck. 
I am consistently getting 12501, even though it allows to pick a account from the list but immediately fails.

**Comment by davyzhang** [comment]
Here's the way I finally made this right after a head scratching day.
1. if you use any kind of framework build like me using ionic to build android version remember to re-sign the apk by https://github.com/noamtamim/apk-resigner/blob/master/signapk.sh , Do NOT directly using jarsigner, because the apk might already signed by debug key

2.update you android sdk to the latest and remove the old one, and if using ionic remember to ionic platform remove android and add it back again to clean the cache
currently the android sdk is 6.0

good luck

thanks @sammuggPS for your great work

I am just on the edge of re-implementing a plugin of my own 😄

**Comment by sammuggPS** [comment]
@davyzhang,
That's awesome! Thank you so much for adding that to this conversation. Hopefully that will help some people out!

**Comment by marianoarga** [comment]
@davyzhang Are you talking about re-signing the debug apk right? or even the PROD build should be re-signed?

**Comment by marianoarga** [comment]
Now getting 12501, let me know if you get any luck @getcarpool @sidharthancrt , etc.

**Comment by davyzhang** [comment]
so far as I can tell, two thing can give 12501 error
1. wrong signature or no signature at all
2. old sdk

but you might check he using the web application id not the oauth client id to get idToken

google has very poor documentations about error codes, :disappointed:

**Comment by getcarpool** [comment]
Sorry I was not able to look into this for quite a while.
@marianoarga , yes I am still getting 12501.
I am using web app client ID and I have re signed the app.
only thing I am yet to do is update my SDK. will let you know the result as soon as I get it done

**Comment by getcarpool** [comment]
Hi @marianoarga, 
I tried to make this work but couldn't. 
so I have switched to the NPM version again. we could get rid of the ugly 12501 but again we were getting error "Failed to retrieve token"
After followed the steps mentioned in issue number 28 to sign the app and installing the signed app (including updating google console with the new signed sha1 key), We are now successfully getting the serverAuthToken.
Hope this helps

**Comment by virenderkverma** [comment]
Hi @marianoarga  @davyzhang @sammuggPS 

My Application was given 12501. After that I follow the steps given by @marianoarga to signed app. After that I getting error 10 ? 

Can anybody help me ?

**Comment by virenderkverma** [comment]
Hi @sammuggPS,

When I am adding latest plugins from github. with client id like this

**cordova plugin add https://github.com/EddyVerbruggen/cordova-plugin-googleplus --variable REVERSED_CLIENT_ID=53379210000-q1pb8a08heitun92b3u5nlh09fkbsole.apps.googleusercontent.com**

without passing scope and webclientid like

 **window.plugins.googleplus.login(
        {
        },
        function (user_data) {
          // For the purpose of this example I will store user data on local storage
          GoogleUserService.setUser({
            userID: user_data.userId,
            name: user_data.displayName,
            email: user_data.email,
            picture: user_data.imageUrl,
            accessToken: user_data.accessToken,
            idToken: user_data.idToken
          });
          $ionicLoading.hide();
           $state.go('tab.chat-detail');
        },
        function (msg) {
          alert(msg);
          $ionicLoading.hide();
        }
      );**

With release version of .apk with signed version. Its working fine on Moto E but not working on Letv mobile. 

In Letv its giving Error code 10.

May I know what is the reason.

**Comment by marianoarga** [comment]
I couldn't manage this to work.
Moved to Facebook Login.
Thank you all for your support.

**Comment by virenderkverma** [comment]
After struggling with this plugin around 2 days . Moved to Oauth based authentication system for google plus. 

Many problem faced with plugins some time conflict with facebook plugins, some time this plugins works and some time not with same device, Plugins not support all devices. Not understand what's issue.

**Comment by sammuggPS** [comment]
Hey all,

Sorry I haven't responded in a while. My responsibilities at work picked up quite a bit recently.

**To those of you using Ionic**, I want you to know that I'm working on a POC to see if I run into any of the same issues. I'll start a separate thread with more info when I get things moving.

Thank you for continuing to talk to one another and collaborating! That's the awesome part about Open Source materials.

~~Sam

**Comment by alexislg2** [comment]
@sammuggPS Any update on this? I'm working on ionic. I removed fully Android SDK then reinstalled Android SDK 6.0. I created Android Key on the Google Developer Api Console using the SHA1 of my debug keystore

But I keep having error 12501 :(

**Comment by sammuggPS** [comment]
@alexislg2 Something that someone pointed out recently is that you need to make sure that you are using an emulator that includes the Google APIs (not all of them do). I updated the troubleshooting section of our documentation to reflect that.

If you are running on a device and still getting the 12501, make sure that you are using a Web Application type clientId. Retrieving an idToken and/or serverAuthCode requires this one, not the one for Android devices.

**Comment by alexislg2** [comment]
thank you for that answer

Yes I'm working on a real device and I use the same idToken I used with [$cordovaOauth](http://ngcordova.com/docs/plugins/oauth/). It's working nice with $cordovaOauth (with awful UX) but got 12501 with your plugin.

**Comment by ramzed** [comment]
Hello, @alexislg2 
I have the same issue. I'm trying to use this plugin instead $cordovaOAuth and also getting error 12501.
I've checked Web Application client id in the Google Console. I'm also working on real device.
Did you make it work?

**Comment by alexislg2** [comment]
No sorry. Glad to know I'm not the only one ;)

**Comment by sammuggPS** [comment]
@alexislg2 and @ramzed,

Can you send me a snippet of how you're calling the login method?
(make sure you sanitize it: block out or replace clientId and other sensitive info)

**Comment by alexislg2** [comment]
Sure:

     $scope.googleLogin = function () {
            $ionicLoading.show({
                template: 'Logging in...'
            });

            window.plugins.googleplus.login(
                {
                    webClientId: settings.GOOGLE_ID, // something like '3746....84-s69aa......8vom...2.apps.googleusercontent.com'
                    scopes: ['https://www.googleapis.com/auth/userinfo.email',
                        'https://www.googleapis.com/auth/userinfo.profile']
                },
                function (user_data) {
                    console.log(user_data);
                    $ionicLoading.hide();
                },
                function (msg) {
                    $ionicLoading.hide();
                    console.error(msg);
                }
            );
        };

**Comment by alexislg2** [comment]
Just a precision I add to change the plugin.xml file so that I can build, like explained here: https://github.com/EddyVerbruggen/cordova-plugin-googleplus/issues/252#issuecomment-229954529

**Comment by sammuggPS** [comment]
1) You don't need to input the profile and email scopes. Those are requested by default.
2) Your scopes are not being input correctly. They should be a space-separated, string list
    i.e. `"<scope1> <scope2>"`
3) Did you install the plugin from npm or from github?

**Comment by matheo** [comment]
Using incorrect scopes gave me 12501 too, even using these: `EMAIL PLUS_LOGIN` which are supposed to be ok

**Comment by sammuggPS** [comment]
@matheo, you should be using the scopes for the Identity Platform (https://developers.google.com/identity/protocols/googlescopes).

Once again, email is requested by default. You do not need to pass that in.

If you want to request PLUS_LOGIN, try `"https://www.googleapis.com/auth/plus.login"`.

**Comment by alexislg2** [comment]
@sammuggPS thank you. I'm so sorry about the scope field, I did not read the doc long enough.

I removed thid field. And now got error 10... I'll have a look at the issues talking about this error

**Comment by alexislg2** [comment]
@sammuggPS for installation I first git cloned the repo in a local directory, then modified the plugin.xml like explained here https://github.com/EddyVerbruggen/cordova-plugin-googleplus/issues/252#issuecomment-229954529 and then `cordova add /path/to/local-repo` so I guess it's equivalent to a git install

**Comment by ramzed** [comment]
I make it works on Android!
First, you need create ID API for OAuth 2.0 Client for android in Google Cloud Console. You need to make sure SHA1 fingerprint of your certificate and package name both have valid values.
Then, you need to create ID API for OAuth 2.0 Client for Web application. 

In my code I'm using client ID from Web application.
`window.plugins.googleplus.login(`
`      {`
`  		  "webClientId": "xxxxxxxx-ggherctrokel7f0s8bpsh1oblepvco75.apps.googleusercontent.com",`
`  	  },`
`      function (user_data) {`
`        $ionicLoading.hide();`
`        console.log(user_data.idToken);`
`        alert(JSON.stringify(user_data));`
`      },`
`      function (msg) {`
`      	console.log(msg);`
`        $ionicLoading.hide();`
`      }`
`    );`

Thank you all for the help and your great work!

**Comment by SidMehta** [comment]
@ramzed : I am doing exactly what you said. Building using phonegap build. :( Still getting 12501 error. @sammuggPS  Just telling the precautions that I followed:
1. Used a "Web Client Id" instead of the android one.
2. Tried putting the android client id json in the project root, no luck.
3. Tried putting the web client id json to project root, no luck. 

Code Snippet for calling the login part:
  window.plugins.googleplus.login
({'webClientId': '26237xxxxxx1-vdgn5xxxxxxxx25aor09u3s14fcicg5a.apps.googleusercontent.com'},
                        function (msg) { }
........ so on..

Config.xml file part:

<gap:plugin spec="https://github.com/EddyVerbruggen/cordova-plugin-googleplus.git" source="git">
</gap:plugin>

Is there anything that I am doing wrong?

**Comment by SidMehta** [comment]
@ramzed: I downloaded the json from here:
![image](https://cloud.githubusercontent.com/assets/6757739/17080204/03197ca8-5144-11e6-8db7-3d726b7a9d00.png)

Do i need to name the file as google-services.json before adding it to the project root?

**Comment by mrfrodo24** [comment]
For those that are still fiddling with the 12501 error that likes to come up when it really shouldn't, something to check might be your **AndroidManifest.xml**.

Specifically, you need to make sure that you have the following line somewhere in it:
`<uses-permission android:name="android.permission.ACCOUNT_MANAGER" />`

**Comment by sammuggPS** [comment]
@SidMehta After working through this with multiple other developers running into the same error, I have come to the following "conclusion":

Error 12501, "user cancelled" (even though the user has not actually cancelled), is generally caused by the wrong SHA1 being used when setting up for Android (either on the Dev Console or [here](https://developers.google.com/mobile/add?platform=android&cntapi=signin)). The SHA1 needs to be the same as the one used when you build your app. I've also noticed that when a developer is using **Ionic** or **Phonegap**, the default keystore is sometimes overridden when using their respective CLIs.

Error 10, "developer error" (another common error when developing for Android), is generally caused by an Android type Client ID being passed into the plugin. The SDK expects a client ID of type Web Application in order to return an idToken and/or serverAuthCode.

Hope this helps.

**Comment by UttamOjha** [comment]
window.plugins.googleplus.login(
        {'scopes':'....'}
 what to pass the values within scope for getting User birthday,currentlocation,and other info.and how to get those details in success callback.
please suggest me or it is better if you have showing me an example.

**Comment by bastaware** [comment]
Ehm... This is probably a bit embarrassing, after struggling with this for a couple of hours I realized that I probably had to run a signed version of the app for the SHA1 to have any effect. And sure enough, with a signed app it's working :)

The weird thing is that I've been using this plugin for long without specifying any webClientId - and I'm sure it worked! So it took me a while to accept that my configuration/code was wrong.

I really urge you to improve the documentation of this plugin + the demo application.  Please make a step by step tutorial (with screenshots) it's not stated very clear or intuitive that a webclientid is needed. Also please consider added comment to the demo app such as /* insert your webclientid here */

**Comment by Wade-McDaniel** [comment]
Hi all,

First, please let me say that all you open-source warriors are my heros!

Down to business: I'm getting the same 12501 response on Android as the others in this thread.

- I'm using Ionic v1 with version ~5.1.1 of this plugin
- I'm passing webClientId:'my-id.apps.googleusercontent.com' and offline: true to the login() method
- I have two keys I've created to try to fix this: an API key in https://console.developers.google.com/apis/credentials associated with my SHA1, and of course the OAuth2 id 'my-id.apps.googleusercontent.com' there as well.
- I'm using Android Studio with an up-to-date SDK (as far as I can tell). Play Services is probably enabled, tho I'm not sure I know how to check that. (It sounds like that shouldn't be a problem tho, right?)

After reading the posts in this thread, I think the google login method in my app is confused about it's permissions, and is bailing out with a (12501) UserCanceled code as for the others here.

**My questions are:**

1. The API key is not necessary, correct? (I was getting the same problem before I created it BTW)
2. How can I best trace where the permissions error is occurring? I'm not seeing a lot of info in the debug console that might illuminate the cause for the permissions error...

Thank you for your help!

**Comment by kidmysoul** [comment]
This is how I solved my 12501 error.

I' using the newest Ionic Framework (platform Android ) in Mac.

1. build your app use `sudo ionic run android`
2. find the apk file and copy it to your jdk/bin forder
   my apk file is in /**yourAppName**/platforms/android/build/outputs/apk 
  and my jdk folder is in /Library/Java/JavaVirtualMachines/jdk1.8.0_77.jdk/Contents/Home/bin/
3. run command `keytool -list -printcert -jarfile apkname.apk` in jdk/bin forder
4.paste the SHA1 value in your google developer console

**Comment by Wade-McDaniel** [comment]
Thanks for replying, kidmysoul!
But I'm not sure that's the right key...

The key that uses the SHA1 value is the API key, and my SHA1 from my APK matches the one in Google Developer Console.

But this thread says to use the **web** key not the API key, right? They key I am supposed to use is from the the "OAuth 2.0 Client IDs "section of this page - https://console.developers.google.com/apis/credentials?project=myproject-1234 - not the one from the "API keys" section, which is where the SHA1 key can be entered. The OAuth 2.0 Client ID doesnt' take SHA1 values, only authorized domains.

1. Do I have the above summary correct? The OAuth key (domain-based) is the one to use, not the API key (SHA1-based), right?
2. Do I have to enter some localhost domain like "http://localhost:port" into the OAuth key to let ionic work, as it's not using a domain name?

**Comment by Wade-McDaniel** [comment]
I was wrong about one part of my understanding: you can enter the SHA1 into an OAuth 2.0 Client ID key if it has a Type=Android. But when I use that key's Client ID as the webClientID I pass to the login() method I still get the UserCanceled error code. So if I use an OAuth 2.0 key where Type="Android" or Type="Web application" I still get the UserCanceled code on Android

Below is the OAuth key for Type=Android (which also doesn't seem to work for me)

![image](https://cloud.githubusercontent.com/assets/17987159/23216371/93ef1548-f8ca-11e6-9df8-e4f7329d13bb.png)

**Comment by kidmysoul** [comment]
>>But when I use that key's Client ID as the webClientID I pass to the login() method I still get the UserCanceled error code. 

1. Use OAuth 2.0 Client ID section clientID which Type=WebApplication as your login webClientID
2. Make your OAuth 2.0 Client ID section which Type= Android SHA-1  matched your debug.keystore's SHA-1.
3. If you still get 12501 error, perhaps the ionic CLI does not use the correct debug key (which should be the key in debug.keystore folder)
so change your Type= Android SHA-1 to the one which is made from apk file as I described before.

**Comment by Wade-McDaniel** [comment]
Thanks for sticking this out with me!

I think it's working for me now, on both iOS and Android. I had been changing too many things trying to get this to work. But now that I know that Android needed two parts: the oAuth type=Android (for app auth) and oAuth type=Web (for later server verification) it seems to be working.

Thanks again!

**Comment by Wade-McDaniel** [comment]
I should add for posterity that I still get the 12501 issue when running on an actual device through Android Studio so I can debug. But when I download the Alpha from Google Play it works, so I don't care much.

**Comment by J-Yen** [comment]
I fixed the problem by specifying the location of the keystore:
https://www.webniraj.com/2016/11/21/cordova-changing-the-default-debug-keystore-for-android-applications/

**Comment by Riyaz0001** [comment]
thanks all, I have solved this problem. error 12501

**Comment by Riyaz0001** [comment]
can anyone tell me, how to get a user birthday in this plugin? how do use scope?

**Comment by whiterook6** [comment]
Thanks for making this plugin. I've gotten it working on IOS and am now struggling with error 12501 while getting it work on android. I've read through the guide and the links posted in the README there are some things that aren't clear to me. Can someone follow my steps and point out what I've done wrong?

1. No certificate by default.
2. Do I need to make a release alias as well as a debug alias? Which SHA-1 do I upload?
3. I've gotten like a million different client IDs. Which do I use?

I'm using phonegap, and haven't opened android studio other than to install google play services and the sdk, so I don't have a certificate by default. I created one using the guide on https://developer.android.com/studio/publish/app-signing.html#signing-manually with the following:

```keytool -genkey -v -keystore my-release-key.jks -keyalg RSA -keysize 2048 -validity 10000 -alias tex_debug_alias```

I then executed the "Get your SHA-1 fingerprint" command:

```
keytool -exportcert -list -v -alias tex_debug_alias -keystore my-keystore.jks
> 25:xxxxxxxxxxxxxxxx:07
```

and saved that SHA-1 in the "Enable Google Services" for android page. This game me a json file that contains this:
```
...
"oauth_client": [
        {
          "client_id": "3xxxxxx7-ixxxxxxxxxxxxxxxxs.apps.googleusercontent.com",
          "client_type": 1,
          "android_info": {
            "package_name": "com.aggregateiq.five",
            "certificate_hash": "6xxxxxxxxxxx8"
          }
        },
        {
          "client_id": "3xxxxxx7-kxxxxxxxxxxu.apps.googleusercontent.com",
          "client_type": 3
        }
      ],
      "api_key": [
        {
          "current_key": "Axxxxxxxxxx6-8xxxxxxxxxxxA"
        }
      ],
      ...
```

I'm writing the app for both IOS and Android, so I made sure to create a plist file for ios. The IOS version works properly. The plist file has a different client id:

```
<key>CLIENT_ID</key>
<string>3xxxxxxx7-cxxxxxxxxxxxxxb.apps.googleusercontent.com</string>
```

There's a web client ID in the Google Developer Credentials page:

```
Web client (auto created by Google Service)	Apr 11, 2017	Web application	3xxxxxxxxx7-kxxxxxxxxxxxxxxxx2.apps.googleusercontent.com	
```

Finally, in the config.xml, I've added the following xml:

```
<gap:plugin name="cordova-plugin-googleplus" source="npm">
  <param name="REVERSED_CLIENT_ID" value="com.googleusercontent.apps.3xxxxxxxxx7-ixxxxxxxxxxxxxxxs" /> // oauth client ID from google services json file.
</gap:plugin>
```

Questions: The guides mentioned `release` vs `debug`, aliases, flavors, etc. Which SHA-1 do I need, and have I created the certificate correctly? Which of the many client ids do I put in the config.xml file--the IOS client id, the IOS reverse client ID, the android client id, or the web client id?

Finally, here's the javascript used to trigger the dialog and that returns 12501:

```
return this.$q(function(resolve, reject){
	window.plugins.googleplus.login({
		"webClientId": "3xxxxxxxxx7-kxxxxxxxxxxxxxxxx2.apps.googleusercontent.com",
	}, resolve, reject);
});

```

Can someone point out what I'm doing wrong?

**Comment by cheukhlee** [comment]
I was having the 12501 error too. One thing that I overlooked was the package name in google developer console not matching the package name in my AndroidManifest.xml file. After updating the package name in console, it started to work for me.

**Comment by lAndresul** [comment]
@kidmysoul Thanks a lot, it was like 16 hours trying to figure it out how fix this! 
It was the wrong sha1 fingerprint, with your method I obtained the correct one.

I promise make a tutorial how to implement all ids and the sha1 correctly.

**Comment by kidmysoul** [comment]
@whiterook6 

the REVERSED_CLIENT_ID in config.xml will be like "***************.apps.googleusercontent.com"

**Comment by nerdmed** [comment]
I really cant believe this. Spend almost 2 full days on this crap. I have tried everything explained here and in the docs. Still getting 12501. Isn't there any other way to debug this issues. Anything we can log on the native side or something. Getting an 12501 for everything that can go wrong is really not helpful.

**Comment by lAndresul** [comment]
@nerdmed may be this can help, my problem was that i was getting the wrong sha1, i was getting the sha1 from android debug apk, but you need the sha1 from the debug apk of ionic.

May be you can try this:

https://stackoverflow.com/questions/11331469/how-do-i-find-out-which-keystore-was-used-to-sign-an-app

**Comment by m4aone** [comment]
oh wow i really got blocked to comment on this issue - puh really not meant to annoy anyone here nothing personal. @nerdmed here. @lAndresul thanks a lot for your answer. I have unzipped the APK i am testing on the device and printed out the SHA1 and its exactly the one that i have inside firebase for the Android App. Any idea how to debug next? I was really hoping the SHA1 would be wrong but, got this right 😢
